File: 2019-01-04/9012/9012.java
import java.util.*;
import java.io.*;

class Main{
	private static Stack <String> stack = new Stack<String>();
	public static void main(String[] args) throws Exception{
		BufferedReader bfr = new BufferedReader(new InputStreamReader(System.in));
		int T = Integer.parseInt(bfr.readLine());

		for(int i = 0; i < T; i++){
			push(bfr.readLine());

			if(isVPS())
				System.out.println("YES");
			else
				System.out.println("NO");

			stack.clear();
		}
	}

	private static void push(String ps){
		for(int i = 0; i < ps.length(); i++)
			stack.push(ps.charAt(i)+"");
	}

	private static Boolean isVPS(){
		char[] ps = new char[52];
		int index = 0;

		while(!stack.empty()){
			
			if(stack.peek().equals("(")){
				if(isNegative(index) || ps[index-1] == '(')
					return false;
				else if(ps[index-1] == ')'){
					index = leftBracket(ps, index);
				}
			}else{
				index = rightBracket(ps, index);
			}
		}
		if(ps[0] == '(' ||ps[0] == ')')
			return false;
		return true;
	}

	private static Boolean isNegative(int index){
		return index-1 < 0;
	}

	private static int leftBracket(char[] ps,int index){
		ps[index-1] = '\0';
		pop();
		index-=1;
		return index;
	}

	private static int rightBracket(char[] ps, int index){
		ps[index] = stack.peek().charAt(0);
		pop();
		index+=1;
		return index;
	}

	private static void pop(){
		stack.pop();
	}

}

File: 2019-01-04/9012/9012.cpp
#include <cstdio>
#include <stack>
#include <string.h>

using namespace std;

stack <char> ps;


void push(char inputPs[]);
int checkYesOrNo();
void stackInit();

int main(){
	int T;
	char inputPs[52];
	scanf("%d", &T);
	for(int i = 0; i < T; i++){
		scanf("%s", &inputPs);
		push(inputPs);
		if(checkYesOrNo())
			printf("YES\n");
		else
			printf("NO\n");
		stackInit();
	}
}	

void stackInit(){
	while(!ps.empty())
		ps.pop();
}

void push(char inputPs[]){
	int length = strlen(inputPs);
	for(int i = 0; i < length; i++)
		ps.push(inputPs[i]);
}


int checkYesOrNo(){
	char receivePs[52];
	int id = 0;
	while(!ps.empty()){
		if(ps.top() == '('){
			if(id-1 <0)
				return 0;
			else if(receivePs[id-1] == ')'){
				receivePs[--id] = '\0';
				ps.pop();
			}
			else
				return 0;
			
		}else{
			receivePs[id] = ps.top();
			id++;
			ps.pop();
		}
	}

	int length = strlen(receivePs);
	if(length > 0)
		return 0;
	return 1;
}